### ISSUE TITLE
Medical Data Visualizer (Certification Project)

### ISSUE THREAD

**Issue by beaucarnes** [freeCodeCamp/CurriculumExpansion#238]
Create project from the Python data analysis certification.

**Comment by scissorsneedfoodtoo** [comment]
Project prototype: https://repl.it/@BeauCarnes/fcc-medical-data-visualizer

**Comment by rlabuonora** [comment]
## Opening paragraph:

>> The dataset values were collected during medical examinations.

I would write a little more about the context of the data: 
 - What do the rows represent? (Patients) 
 - What information do the columns contains (patient's body measurements, results from various blood tests, and lifestyle choices).

## Questions

>> Convert the data into long format and create a chart that shows the value counts of the categorical features using seaborn's catplot(). The dataset should be split by 'Cardio' so there is one chart for each 'cardio' value. The chart should look like "examples/Figure_1.png".

- The stub for this function in `medical_data_visualizer.py` is `draw_cat_plot`, but `main.py` calls `medical_data_visualizer.draw_factor_plot()`.

- This is a long problem, and it's mostly about creating the plot, so I would highlight that on the wording of the question. I would describe the desired visualization first, and then outline the steps required to build it.

    - Create a chart similar to`examples/Figure_1.png`, where we show the counts of good and bad outcomes for (ac, alco, active, cholesterol, gluc, smoke) for patients with cardio=1 and cardio=0 in different panels.

I tried to follow the instructions in the code comments but got stuck. Instead, I tried to replicate the plot in the example using this:

    - https://stackoverflow.com/questions/35692781/python-plotting-percentage-in-seaborn-bar-plot 
    - https://seaborn.pydata.org/generated/seaborn.catplot.html

Here is my solution, which produces the visualization and is a little cleaner than the proposed solution (I think).

```
  medical_dict = { 1: 0, 2 : 1, 3: 1}

  df['cholesterol'] = df['cholesterol'].map( medical_dict )

  df['gluc'] = df['gluc'].map( medical_dict ) 

  df_cat = df.groupby(["cardio", "cholesterol", "gluc", "alco", "active"]).size().rename("total").reset_index().melt(['total', 'cardio'])

  sns.catplot(data = df_cat, x='variable', y='total', hue='value', col='cardio', kind="bar")

  plt.savefig('catplot.png')

```
The resulting plot looks like the solution but does not pass the tests:
![catplot](https://user-images.githubusercontent.com/10402561/72672760-2583cf00-3a3e-11ea-8fb3-e70128eb0f58.png)

**Comment by rayjohnson529** [comment]
I posted this in the wrong place initially:

The solution doesn't seem to result in the figure posted.

I am going to submit my results soon, but this seems like maybe an earlier draft of the starter problem?

I peaked at the solution to try to see how mine compared and copied the code and the resulting graph didn't match 'figure 1' - I believe the problem has something to do with the part using "fig, ax" (unless somehow I've made a major mistake).

In addition, the function in the solution doesn't match the name of the function in the problem provided (it changed from 'draw_factor_plot' to 'draw_cat_plot.')

The instructions call for us to include the 'overweight' column we created, but 'Figure_1' does not include the 'overweight' column.

The test also calls for 'draw_factor_plot' - perhaps this is all on me...
Will do a more thorough write up once challenge is completed.

**Comment by scissorsneedfoodtoo** [comment]
Thank you for your detailed reviews @rlabuonora and @rayjohnson529. The `draw_cat_plot`/`draw_factor_plot` is something we forgot to update in the boilerplate. Also good catch with the wording and `df['ap_lo'] <= df['ap_hi'])` -- we will have to check the graphs and adjust either the wording or the snippet itself. Also, you're right that `Figure_1` doesn't include the overweight column, but `Figure_2` does.

@rlabuonora, which syntax errors are you getting? Could you post a link to your solution on Repl.it?

**Comment by rlabuonora** [comment]
>> @rlabuonora, which syntax errors are you getting? Could you post a link to your solution on Repl.it?

I meant the starter code raises syntax errors as is, That's probably because of lines like:

```
fig, ax = None
sns.catplot(     ) # fill in the parenthesis
```
This means that you can't use `main.py` as is for experimenting without commenting out the line that imports the solution.

**Comment by scissorsneedfoodtoo** [comment]
@rlabuonora, I see, that's a good point. Thank you for clarifying.

What we could do is have those bits of unfinished code commented out and change the comments to read, "Uncomment and fill in the parenthesis".

**Comment by rlabuonora** [comment]
Yes, I think that would work.

**Comment by beaucarnes** [comment]
I implemented a lot of the suggestions. 

@rlabuonora You had some very helpful advice. It would be great if you had a chance to review this again with the changes. I tried to get your solution code to work but it was changing the value of "total". In the original code, the highest total value is around 30K but in your code it is around 10K. The truth is I'm kind of new to pandas myself so I wasn't able to figure out the reason for the discrepancy. While I implemented much of what you suggested, some of the things I couldn't figure out the best way to implement. I'm definitely open to more suggestions. If you want to make your own version of the challenge with tests and description, we may be able to use it.
